vsintegration/src/FSharp.Editor/Completion/CompletionUtils.fs

```diff
@@ -82,18 +82,21 @@ module internal CompletionUtils =
     let isStartingNewWord (sourceText, position) =
         CommonCompletionUtilities.IsStartingNewWord(sourceText, position, (fun ch -> isIdentifierStartCharacter ch), (fun ch -> isIdentifierPartCharacter ch))
 
-    let shouldProvideCompletion (documentId: DocumentId, filePath: string, defines: string list, text: SourceText, position: int) : bool =
-        let textLines = text.Lines
-        let triggerLine = textLines.GetLineFromPosition position
-        let colorizationData = Tokenizer.getColorizationData(documentId, text, triggerLine.Span, Some filePath, defines, CancellationToken.None)
+    let shouldProvideCompletion (documentId: DocumentId, filePath: string, defines: string list, sourceText: SourceText, triggerPosition: int, triggerChar: char) : bool =
+        let textLines = sourceText.Lines
+        let triggerLine = textLines.GetLineFromPosition triggerPosition
+        let colorizationData = Tokenizer.getColorizationData(documentId, sourceText, triggerLine.Span, Some filePath, defines, CancellationToken.None)
         colorizationData.Count = 0 || // we should provide completion at the start of empty line, where there are no tokens at all
         colorizationData.Exists (fun classifiedSpan -> 
-            classifiedSpan.TextSpan.IntersectsWith position &&
+            classifiedSpan.TextSpan.IntersectsWith triggerPosition &&
             (
                 match classifiedSpan.ClassificationType with
                 | ClassificationTypeNames.Comment
                 | ClassificationTypeNames.StringLiteral
                 | ClassificationTypeNames.ExcludedCode
+                | ClassificationTypeNames.Operator
                 | ClassificationTypeNames.NumericLiteral -> false
                 | _ -> true // anything else is a valid classification type
-            ))
\ No newline at end of file
+            ))
+        &&
+        (triggerChar = '.' || (Settings.IntelliSense.ShowAfterCharIsTyped && isStartingNewWord(sourceText, triggerPosition)))
```

Comment:
I understand the logic of putting this here. But, considering the comment I made on ShouldTriggerCompletion and ProvideCompletionsAsync, this bit that's checking does not belong in the flow of ProvideCompletionsAsync. I recommend putting this back to the way it was.

Reply:
Ah, I see. Hmmm. This logic is really just strewn about. I'd like to clean it up in this PR.

Reply:
If you want to clean things up, just create some more functions that contain these kind of checks. :)

Reply:
Yeah, I'll just do that separately. I'm thinking of a winter project of porting the old completion tests over to CompletionProvider so I might as well try to clean a few things up at this time too